code/controllers/subsystem/minimaps.dm

```diff
@@ -0,0 +1,281 @@
+	///Actual icon of the drawn zlevel with all of it's atoms
+	var/icon/hud_image
+	var/list/xeno_images = list()
+	var/list/xeno_images_raw = list()
+	var/list/marine_images = list()
+	var/list/marine_images_raw = list()
+	var/list/alpha_images = list()
+	var/list/alpha_images_raw = list()
+	var/list/bravo_images = list()
+	var/list/bravo_images_raw = list()
+	var/list/charlie_images = list()
+	var/list/charlie_images_raw = list()
+	var/list/delta_images = list()
+	var/list/delta_images_raw = list()
+
+/**
+ * Holder datum to ease updating of atoms to update
+ */
+/datum/minimap_updator
+	/// Atom to update with the overlays
+	var/atom/minimap
+	///Target zlevel we want to be updating to
+	var/ztarget = 0
+
+/datum/minimap_updator/New(minimap, ztarget)
+	..()
+	src.minimap = minimap
+	src.ztarget = ztarget
+
+/**
+ * Adds an atom we want to track with blips to the subsystem
+ * Arguments:
+ * * target: atom we want to track
+ * * zlevel: zlevel we want thhis atom to be tracked for
+ * * hud_flags: tracked HUDs we want this atom to be displayed on
+ * * iconstate: iconstate for the blip we want to be used for this tracked atom
+ * * icon: icon file we want to use for this blip, 'icons/UI_icons/map_blips.dmi' by efault
+ */
+/datum/controller/subsystem/minimaps/proc/add_marker(atom/target, zlevel, hud_flags = NONE, iconstate, icon = 'icons/UI_icons/map_blips.dmi')
+	if(!isatom(target) || !zlevel || !hud_flags || !iconstate || !icon)
+		CRASH("Invalid marker added to subsystem")
+	if(!initialized)
+		earlyadds += CALLBACK(src, .proc/add_marker, target, zlevel, hud_flags, iconstate, icon)
+		return
+	var/image/blip = image(icon, iconstate, pixel_x = MINIMAP_PIXEL_FROM_WORLD(target.x), pixel_y = MINIMAP_PIXEL_FROM_WORLD(target.y))
+	images_by_source[target] = blip
+	if(hud_flags & MINIMAP_FLAG_XENO)
+		minimaps_by_z["[zlevel]"].xeno_images[target] = blip
+		minimaps_by_z["[zlevel]"].xeno_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_MARINE)
+		minimaps_by_z["[zlevel]"].marine_images[target] = blip
+		minimaps_by_z["[zlevel]"].marine_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_ALPHA)
+		minimaps_by_z["[zlevel]"].alpha_images[target] = blip
+		minimaps_by_z["[zlevel]"].alpha_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_BRAVO)
+		minimaps_by_z["[zlevel]"].bravo_images[target] = blip
+		minimaps_by_z["[zlevel]"].bravo_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_CHARLIE)
+		minimaps_by_z["[zlevel]"].charlie_images[target] = blip
+		minimaps_by_z["[zlevel]"].charlie_images_raw += blip
+	if(hud_flags & MINIMAP_FLAG_DELTA)
+		minimaps_by_z["[zlevel]"].delta_images[target] = blip
+		minimaps_by_z["[zlevel]"].delta_images_raw += blip
+	if(ismovableatom(target))
+		RegisterSignal(target, COMSIG_MOVABLE_Z_CHANGED, .proc/on_z_change)
+		RegisterSignal(target, COMSIG_MOVABLE_MOVED, .proc/on_move)
+	removal_cbs[target] = CALLBACK(src, .proc/removeimage, blip, target)
+	RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/remove_marker)
```

Comment:
This code is for removing the marker only on parent delet8ion to prevent refs, but you could make it happen at a different point yes, though I want to see what happens in TM before I decide